examples/UNet/massachussets_road_segm.py

```diff
@@ -0,0 +1,340 @@
+    images_map = [img for img in os.listdir(os.path.join(folder, "map")) if fnmatch.fnmatch(img, "*.tif*")]
+    assert(len(images_sat) == len(images_map))
+    images_sat.sort()
+    images_map.sort()
+    # images are 1500 by 1500 pixels each
+    data = np.zeros((len(images_sat), 3, 1500, 1500), dtype=np.uint8)
+    target = np.zeros((len(images_sat), 1, 1500, 1500), dtype=np.uint8)
+    ctr = 0
+    for sat_im, map_im in zip(images_sat, images_map):
+        data[ctr] = plt.imread(os.path.join(folder, "sat_img", sat_im)).transpose((2, 0, 1))
+        # target has values 0 and 255. make that 0 and 1
+        target[ctr, 0] = plt.imread(os.path.join(folder, "map", map_im))/255
+        ctr += 1
+    return data, target
+
+def batch_generator(data, target, BATCH_SIZE):
+    '''
+    just a simple batch iterator, no cropping, no rotation, no anything
+    '''
+    np.random.seed()
+    idx = np.arange(data.shape[0])
+    while True:
+        ids = np.random.choice(idx, BATCH_SIZE)
+        yield np.array(data[ids]), np.array(target[ids])
+
+def random_crop_generator(generator, crop_size=(128, 128)):
+    '''
+    yields a random crop of size crop_size
+    '''
+    np.random.seed()
+    if type(crop_size) not in (tuple, list):
+        crop_size = [crop_size, crop_size]
+    elif len(crop_size) == 2:
+        crop_size = list(crop_size)
+    else:
+        raise ValueError("invalid crop_size")
+    for data, seg in generator:
+        lb_x = np.random.randint(0, data.shape[2]-crop_size[0])
+        lb_y = np.random.randint(0, data.shape[3]-crop_size[1])
+        data = data[:, :, lb_x:lb_x+crop_size[0], lb_y:lb_y+crop_size[1]]
+        seg = seg[:, :, lb_x:lb_x+crop_size[0], lb_y:lb_y+crop_size[1]]
+        yield data, seg
+
+def threaded_generator(generator, num_cached=10):
+    # this code is written by jan Schluter
+    # copied from https://github.com/benanne/Lasagne/issues/12
+    import Queue
+    queue = Queue.Queue(maxsize=num_cached)
+    sentinel = object()  # guaranteed unique reference
+
+    # define producer (putting items into queue)
+    def producer():
+        for item in generator:
+            queue.put(item)
+        queue.put(sentinel)
+
+    # start producer (in a background thread)
+    import threading
+    thread = threading.Thread(target=producer)
+    thread.daemon = True
+    thread.start()
+
+    # run as consumer (read items from queue, in current thread)
+    item = queue.get()
+    while item is not sentinel:
+        yield item
+        queue.task_done()
+        item = queue.get()
+
+def prepare_dataset():
+    prep_folders()
+    prep_urls()
+    download_dataset()
+    # the dataset is now downloaded in the background. Once every few seconds we check if the download is done. We do
+    # this by checking whether tha last training image exists
+    while not os.path.isfile("data/training/map/99238675_15.tiff") and not os.path.isfile("data/training/sat_img/99238675_15.tiff"):
```

Comment:
You can't be 100% sure that the last training image is the last one to be downloaded, though -- if everything happens in parallel, one of the shorter scripts might still end later than the longest script.
Anyway, if you use `map()` as suggested above, you won't need to poll.